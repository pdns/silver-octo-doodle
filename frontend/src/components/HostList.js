import React from 'react';
import PropTypes from 'prop-types';
import List from '@material-ui/core/List';
import Dialog from '@material-ui/core/Dialog';
import Divider from '@material-ui/core/Divider';
import ListItem from '@material-ui/core/ListItem';
import DialogTitle from '@material-ui/core/DialogTitle';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import DialogContent from '@material-ui/core/DialogContent';
import AddIcon from '@material-ui/icons/Add';
import { graphql, createFragmentContainer } from 'react-relay';
import HostListItem from './HostListItem';
import HostInputForm from './HostInputForm';
import CreateHostMutation from '../mutations/CreateHostMutation';

class HostList extends React.Component {
  constructor(props) {
    super(props);
    this.state = { dialog: false };
  }

  handleClick = () => {
    this.setState({ dialog: true });
  }

  handleClose = () => {
    this.setState({ dialog: false });
  }

  handleCreate = (input) => {
    CreateHostMutation.commit(this.props.relay.environment, input);
    this.handleClose();
  }

  render() {
    const hostsItems = this.props.userHostData.hosts.edges.map(e =>
      <HostListItem key={e.node.id} host={e.node} />);

    return (
      <React.Fragment>
        <ListItem button onClick={this.handleClick}>
          <ListItemIcon>
            <AddIcon />
          </ListItemIcon>
          <ListItemText primary="Create host" />
        </ListItem>
        <Divider />
        <Dialog open={this.state.dialog} fullWidth>
          <DialogTitle>
            Create new host
          </DialogTitle>
          <DialogContent>
            <HostInputForm onCancel={this.handleClose} onSubmit={this.handleCreate} />
          </DialogContent>
        </Dialog>
        <List>
          {hostsItems}
        </List>
      </React.Fragment>
    );
  }
}

HostList.propTypes = {
  userHostData: PropTypes.object.isRequired,
  relay: PropTypes.any,
};

export default createFragmentContainer(
  HostList,
  graphql`
    fragment HostList_userHostData on User {
      hosts(first: 2147483647) @connection(key: "HostList_hosts") {
        edges {
          node {
            id
            ...HostListItem_host
          }
        }
      }
    }
  `,
);
