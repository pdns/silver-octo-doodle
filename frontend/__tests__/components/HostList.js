/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import HostList from '../../src/components/HostList';

jest.mock('react-relay', () => ({
  createFragmentContainer: x => x,
}));

jest.mock('@material-ui/core/List', () => 'List');
jest.mock('@material-ui/core/Dialog', () => 'Dialog');
jest.mock('@material-ui/core/Divider', () => 'Divider');
jest.mock('@material-ui/core/ListItem', () => 'ListItem');
jest.mock('@material-ui/core/DialogTitle', () => 'DialogTitle');
jest.mock('@material-ui/core/ListItemIcon', () => 'ListItemIcon');
jest.mock('@material-ui/core/ListItemText', () => 'ListItemText');
jest.mock('@material-ui/core/DialogContent', () => 'DialogContent');
jest.mock('@material-ui/icons/Add', () => 'AddIcon');

jest.mock('../../src/components/HostListItem', () => 'HostListItem');
jest.mock('../../src/components/HostInputForm', () => 'HostInputForm');
jest.mock('../../src/mutations/CreateHostMutation', () => ({ commit: () => ({}) }));

describe('HostList', () => {
  it('should render correctly with no hosts', () => {
    const data = {
      hosts: {
        edges: [],
      },
    };
    const relay = {
      environment: 'relay-env',
    };
    const tree = renderer
      .create(<HostList userHostData={data} relay={relay} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with hosts', () => {
    const data = {
      hosts: {
        edges: [
          { node: { id: '1' } },
        ],
      },
    };
    const relay = {
      environment: 'relay-env',
    };
    const tree = renderer
      .create(<HostList userHostData={data} relay={relay} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
