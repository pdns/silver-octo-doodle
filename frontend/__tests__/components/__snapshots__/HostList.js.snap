// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HostList should render correctly with hosts 1`] = `
Array [
  <ListItem
    button={true}
    onClick={[Function]}
  >
    <ListItemIcon>
      <AddIcon />
    </ListItemIcon>
    <ListItemText
      primary="Create host"
    />
  </ListItem>,
  <Divider />,
  <Dialog
    fullWidth={true}
    open={false}
  >
    <DialogTitle>
      Create new host
    </DialogTitle>
    <DialogContent>
      <HostInputForm
        onCancel={[Function]}
        onSubmit={[Function]}
      />
    </DialogContent>
  </Dialog>,
  <List>
    <HostListItem
      host={
        Object {
          "id": "1",
        }
      }
    />
  </List>,
]
`;

exports[`HostList should render correctly with no hosts 1`] = `
Array [
  <ListItem
    button={true}
    onClick={[Function]}
  >
    <ListItemIcon>
      <AddIcon />
    </ListItemIcon>
    <ListItemText
      primary="Create host"
    />
  </ListItem>,
  <Divider />,
  <Dialog
    fullWidth={true}
    open={false}
  >
    <DialogTitle>
      Create new host
    </DialogTitle>
    <DialogContent>
      <HostInputForm
        onCancel={[Function]}
        onSubmit={[Function]}
      />
    </DialogContent>
  </Dialog>,
  <List />,
]
`;
